reference/mongodb/bson/iterator/next.xml

```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<!-- $Revision$ -->
+<refentry xml:id="mongodb-bson-iterator.next" xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink">
+ <refnamediv>
+  <refname>MongoDB\BSON\Iterator::next</refname>
+  <refpurpose>Advances the iterator to next element</refpurpose>
+ </refnamediv>
+
+ <refsect1 role="description">
+  &reftitle.description;
+  <methodsynopsis>
+   <modifier>public</modifier> <type>void</type><methodname>MongoDB\BSON\Iterator::next</methodname>
```

Comment:
I realize the return type is `void`, but when checking the implementation I realized it unconditionally increments `intern->key`:

```c
intern->valid = bson_iter_next(&intern->iter);
intern->key++;
php_phongo_iterator_free_current(intern);
```

Does it make sense to do so if the position is not valid? I suppose once we reach an invalid position, we'd never get back to a valid position where `Iterator::key()` could be called successfully -- so maybe this isn't a concern.

Reply:
Interesting find, and a good point. This could be a problem if we don't properly handle rewinding the iterator, but then `next` might as well guard against an invalid iterator and no-op if the iterator is already invalid. While that strategy could then cause unexpected behaviour if the underlying `bson_t` were to change (which currently isn't possible), this is already an issue in libbson and explained in the `bson_iter_t` docs:
> The [bson_t](http://mongoc.org/libbson/current/bson_t.html) MUST be valid for the lifetime of the iter and it is an error to modify the [bson_t](http://mongoc.org/libbson/current/bson_t.html) while using the iter.

I'll add some tests in https://github.com/mongodb/mongo-php-driver/pull/1435.